File: dsa/week(5)/stl1.js
// let v1 = new Array(5);
// for (let i = 0; i < 5; i++) {
//   const readline = require("readline").createInterface({
//     input: process.stdin,
//     output: process.stdout,
//   });
//   readline.question(`Enter element ${i + 1}: `, (input) => {
//     v1[i] = parseInt(input);
//     if (i === 4) {
//       readline.close();
//       main(v1);
//     }
//   });
// }

// function main(v1) {
//   console.log(v1.length);
//   v1.push(-1);
//   v1.pop();

//   // Clear the array
//   v1.length = 0;
//   console.log(v1[v1.length - 1]); // equivalent to v1.back()
//   console.log(v1[0]); // equivalent to v1.front()
// }

// let v1 = [1, 2, 3, 4, 5];
// for (let i = 0; i < v1.length; i++) {
//   console.log(v1[i]);
// }
// for (let ele of v1) {
//   process.stdout.write(ele + " "); // Use process.stdout.write to avoid newline
// }
// console.log(); // Newline after printing all elements
// // Check if the array is empty
// console.log(v1.length === 0);

// let v1 = new Array(10).fill(1);
// v1 = v1.reverse();
// v1.splice(1, 3);
// v1.sort((a, b) => b - a);
// [v1[0], v1[1]] = [v1[1], v1[0]];

// let v1 = [];
// for (let i = 0; i < 5; i++) {
//   v1.push([]);
// }

// // Resize each sub-array to have 'c' elements, initialized to 0
// for (let i = 0; i < 5; i++) {
//   for (let j = 0; j < c; j++) {
//     v1[i].push(0);
//   }
// }

// Create a 1D array filled with 1s, equivalent to C++ vector<int> v2(5, 1);
// let v2 = new Array(5).fill(1);

// const readline = require("readline");
// const rl = readline.createInterface({
//   input: process.stdin,
//   output: process.stdout,
// });

// rl.question("Enter row and column (space-separated): ", (input) => {
//   const [row, col] = input.split(" ").map(Number);
//   let v3 = Array(row).fill().map(() => Array(col).fill(0)); // vector<vector<int>> v3(row, vector<int>(col));
//   console.log(v3);
//   rl.close();
// });

// let students = [];
// let student = {
//   id: id,
//   grade: grade,
// };
// function addStudent(id, grade) {
//   let student = {
//     id: id,
//     grade: grade,
//   };
//   students.push(student);
// }

// addStudent(1, "A");
// addStudent(2, "B");
// addStudent(3, "C");
// console.log(students);

// Pair
// let p1 = {};
// let p3 = { first: 10, second: 'A' };
// function makePair(first, second) {
//     return { first, second };
// }
// let p9 = makePair(10, 'A');

const readline = require("readline").createInterface({
  input: process.stdin,
  output: process.stdout,
});
let students = [];
let num;

// Function to ask question and get input from user
function askQuestion(query) {
  return new Promise((resolve) =>
    readline.question(query, (ans) => resolve(ans))
  );
}

(async () => {
  num = parseInt(await askQuestion("Enter number of students: "));

  for (let i = 0; i < num; i++) {
    let roll_number = await askQuestion("Enter roll number: ");
    let section = await askQuestion("Enter section: ");
    students.push({ roll_number, section });
  }

  // Sorting the students array
  students.sort((a, b) => {
    // Assuming the comparison logic is based on roll_number
    return a.section - b.section;
  });

  // Printing the sorted list of students
  for (let i = 0; i < num; i++) {
    console.log(`${students[i].roll_number} ${students[i].section}`);
  }
  readline.close();
})();

File: dsa/week(5)/stl1.py
from operator import itemgetter

def comp(a, b):
    return a[1] > b[1]

def main():
    # Vector equivalent in Python: List
    v1 = []
    for i in range(5):
        v1.append(int(input()))

    print(len(v1))  # Size of the list
    print(len(v1))  # Capacity doesn't apply in Python; len(v1) represents the current size

    v1.append(-1)   # Push back
    v1.pop()        # Pop back
    v1.clear()      # Clear the list
    # v1.back() and v1.front() do not exist in Python but can be accessed via v1[-1] and v1[0]

    # Iterating with iterators and auto-equivalent in Python
    for ele in v1:
        print(ele)

    print(int(len(v1) == 0))  # Equivalent of v1.empty() in C++

    # `memset` has no direct equivalent in Python, use list comprehension
    v1 = [1] * 5  # Filling the list with 1s
    # Reverse a portion of the list
    v1[1:4] = reversed(v1[1:4])
    # Erase elements from 1 to 3 (inclusive in C++, exclusive in Python)
    del v1[1:4]

    # Sorting in reverse order
    v1.sort(reverse=True)

    # `swap` equivalent
    v1[0], v1[1] = v1[1], v1[0]

    # Resize isn't applicable in Python. Use list slicing or dynamic list manipulation

    # 2D vector equivalent
    row, col = map(int, input().split())
    v3 = [[0] * col for _ in range(row)]

    # Pair equivalent: Tuple
    students = []
    num = int(input())
    for _ in range(num):
        roll_number, section = input().split()
        roll_number = int(roll_number)
        students.append((roll_number, section))

    # Sort the list of tuples
    students.sort(key=itemgetter(1), reverse=True)

    for student in students:
        print(student[0], student[1])

if __name__ == "__main__":
    main()
